lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs

```diff
@@ -199,13 +243,98 @@ data SelectionError
     | SelectionOutputsError ErrPrepareOutputs
     deriving (Eq, Show)
 
+-- | Accounts for pre-existing inputs provided to coin selection.
+--
+accountForExistingInputs :: Functor m => ModifySelection m change
+accountForExistingInputs performSelectionFn constraints params =
+    fmap modifyResult <$> performSelectionFn
+        (modifyConstraints constraints)
+        (modifyParams params)
+  where
+    modifyConstraints :: SelectionConstraints -> SelectionConstraints
+    modifyConstraints
+        = over #computeMinimumCost
+            modifyComputeMinimumCost
+        . over #computeSelectionLimit
+            modifyComputeSelectionLimit
+      where
+        modifyComputeMinimumCost
+            :: (SelectionSkeleton -> Coin)
+            -> (SelectionSkeleton -> Coin)
+        modifyComputeMinimumCost = (. modifySkeleton)
+          where
+            modifySkeleton :: SelectionSkeleton -> SelectionSkeleton
+            modifySkeleton =
+                over #skeletonInputCount (+ UTxO.size existingInputs)
+
+        modifyComputeSelectionLimit
+            :: ([TxOut] -> SelectionLimit)
+            -> ([TxOut] -> SelectionLimit)
+        modifyComputeSelectionLimit = (modifySelectionLimit .)
+          where
+            modifySelectionLimit :: SelectionLimit -> SelectionLimit
+            modifySelectionLimit = fmap $ subtract $ UTxO.size existingInputs
+
+    modifyParams :: SelectionParams -> SelectionParams
+    modifyParams
+        = over #assetsToMint
+            modifyAssetsToMint
+        . over #existingInputs
+            modifyExistingInputs
+        . over #rewardWithdrawal
+            modifyRewardWithdrawal
+        . over #utxoAvailable
+            modifyUTxOAvailable
+      where
+        modifyAssetsToMint :: TokenMap -> TokenMap
+        modifyAssetsToMint = TokenMap.add (view #tokens existingInputValue)
+
+        modifyExistingInputs :: UTxO -> UTxO
+        modifyExistingInputs = const UTxO.empty
+
+        modifyRewardWithdrawal :: Maybe Coin -> Maybe Coin
+        modifyRewardWithdrawal =
+            Just . (view #coin existingInputValue <>) . fromMaybe mempty
+
+        modifyUTxOAvailable :: UTxOIndex -> UTxOIndex
+        modifyUTxOAvailable = UTxOIndex.deleteMany $ UTxO.dom existingInputs
```

Comment:
Oh, didn't think about that, an existing input means that we've effectively lost that input from our UTxO set.

Reply:
> Oh, didn't think about that, an existing input means that we've effectively lost that input from our UTxO set.

Yes that's right! We definitely shouldn't reselect it.

By removing it here, it means that we don't have to modify the internal algorithm: the internal algorithm can assume that everything in `availableUTxO` really is available for selection.